qiskit/circuit/add_control.py

```diff
@@ -0,0 +1,223 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Add control to operation if supported.
+"""
+from qiskit import QiskitError
+from qiskit.extensions import UnitaryGate
+
+
+def add_control(operation, num_ctrl_qubits, label):
+    """Add num_ctrl_qubits controls to operation
+
+    Args:
+        operation (Gate or ControlledGate): operation to add control to.
+        num_ctrl_qubits (int): number of controls to add to gate (default=1)
+        label (str): optional gate label
+
+    Returns:
+        ControlledGate: controlled version of gate. This default algorithm
+            uses num_ctrl_qubits-1 ancillae qubits so returns a gate of size
+            num_qubits + 2*num_ctrl_qubits - 1.
+    """
+    if isinstance(operation, UnitaryGate):
+        # attempt decomposition
+        operation._define()
+    if _control_definition_known(operation, num_ctrl_qubits):
+        return _q_if_predefined(operation, num_ctrl_qubits)
+    return q_if(operation, num_ctrl_qubits=num_ctrl_qubits, label=label)
+
+
+def _control_definition_known(operation, num_ctrl_qubits):
+    if num_ctrl_qubits == 2 and operation.name == 'x':
+        return True
+    elif num_ctrl_qubits == 1:
+        return operation.name in {'x', 'y', 'z', 'h', 'rz', 'swap', 'u1', 'u3', 'cx'}
+    else:
+        return False
+
+
+def _q_if_predefined(operation, num_ctrl_qubits):
+    """Returns controlled gates with hard-coded definitions in
+    the standard extensions."""
+    if operation.name == 'x' and num_ctrl_qubits in [1, 2]:
```

Comment:
I thinks this ifcase should be based on the type, not on the name.

Reply:
yes, that's probably better